packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -85,35 +85,29 @@ export function esbuildDepPlugin(
         }
       )
 
-      function resolveEntry(id: string, isEntry: boolean, resolveDir: string) {
```

Comment:
`resolveDir` was added here to solve #2975, in this PR https://github.com/vitejs/vite/pull/3003#issue-615961395

Would you check that this is no longer an issue after removing it? It would be great to add a test based on #2975 to your growing deps test suite (thanks for that by the way!)

Reply:
`resolveDir` is removed on purpose.

I can describe the history of this bug.

#2975 and #3254 targets to same bug introduced in https://github.com/vitejs/vite/commit/8e1d3d82f6d3ba9e48adb60a1b016c35f0aab627.

The original purpose of https://github.com/vitejs/vite/commit/8e1d3d82f6d3ba9e48adb60a1b016c35f0aab627 is to aviod duplicated modules by separate entry imports from none-entry imports. 

It introduced first version of `resolveEntry` util function:

```js
function resolveEntry(id: string, isEntry: boolean) {
        const flatId = flattenId(id)
        if (flatId in qualified) {
          return isEntry
            ? {
                path: flatId,
                namespace: 'dep'
              }
            : {
                path: path.resolve(qualified[flatId])
              }
        }
      }
```

And we knows for none-entry import, using `qualified[flatId]` would make nested dependency like `node_modules/package-a/node_modules/package-b` never be resolved now.

Then #3003 comes to fix. It uses nodejs api `require.resolve` and resolveDir (provided by esbuild) to resolve none-entry import from sub node_modules directory.

`resolveDir` introduced in #3003:

```js
      function resolveEntry(id: string, isEntry: boolean, resolveDir: string) {
        const flatId = flattenId(id)
        if (flatId in qualified) {
          return isEntry
            ? {
                path: flatId,
                namespace: 'dep'
              }
            : {
                path: require.resolve(flatId, {
                  paths: [resolveDir]
                })
              }
        }
      }
```

#3003 has two problems. The more obvious one is `flatId` + `require.resolve` = module not found. For scoped packages like `@some/package`, `flatId` would be like `some_package`. And nodejs's `require.resolve` can't resolve the flatted id.

So we have third version of `resolveEntry` in #3753:

```js
      function resolveEntry(id: string, isEntry: boolean, resolveDir: string) {
        const flatId = flattenId(id)
        if (flatId in qualified) {
          return isEntry
            ? {
                path: flatId,
                namespace: 'dep'
              }
            : {
                path: require.resolve(qualified[flatId], { // flatId -> qualified[flatId]
                  paths: [resolveDir]
                })
              }
        }
      }
```

#3753 shouldn't be merged, because it breaks fix of #3003. It resolved scoped package problem, but `qualified[flatId]` is a absolute path point to `node_modules/xx/main-of-xx`. The `require.resolve` and `resolveDir` are pointless.

Then I comes in #3753:

```js
      function resolveEntry(id: string, isEntry: boolean, resolveDir: string) {
        const flatId = flattenId(id)
        if (flatId in qualified) {
          return isEntry
            ? {
                path: flatId,
                namespace: 'dep'
              }
            : {
                path: require.resolve(id, { // qualified[flatId] -> id
                  paths: [resolveDir]
                })
              }
        }
      }
```

It looks good, but in `2.4.0-beta.0` beta release we got error reports from #4005 and #4014. The problem is `require.resolve` behaves different with vite's module resolve function. For packages provide es module ("module" in package.json), it still resolve to "main". It's actually the other one problem introduced in #3003.

And as you see this(#4091) is the most recent fix on the long living bug. As described, I think it's not necessary to process none-entry imports in `resolveEntry`, so resolveDir is removed.

What a long story= =

Reply:
And the first version of test case(`test-package-a` and `test-package-b`) is very suit to the circumstance of #2975, so I don't think we needs another test case for this.

Reply:
的确,这段代码是没有任何意义的，现在有没有好的修复方案
Indeed, this code is meaningless. Is there a good fix now

{
                path: require.resolve(id, { // qualified[flatId] -> id
                  paths: [resolveDir]
                })
              }

Reply:
Indeed, this code is meaningless. Is there a good fix now

```js
{
    path: require.resolve(id, { // qualified[flatId] -> id
      paths: [resolveDir]
    })
  }
```

Reply:
Too many issues.

I mean #3053 on third version.

Reply:
Thanks for taking the time to write down the detailed history of the issue. This is very helpful.